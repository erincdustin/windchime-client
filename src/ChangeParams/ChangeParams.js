import React from 'react';

function ChangeParams(props) {

    const options = [
      '',
      .1,
      .2,
      .3,
      .4,
      .5,
      .6,
      .7,
      .8,
      .9,
      1
    ];

    //DRY up mapped parameter functions!
    const mappedTargetEnergy = options.map((energyOption, index) => {
      if (energyOption === '') {
        return <option key={index} value={''}>n/a</option>
      }
      return(
        <option key={index} value={energyOption}>{energyOption * 10}</option>
      )
    })
  
    const mappedTargetValence = options.map((valenceOption, index) => {
      if (valenceOption === '') {
        return <option key={index} value={''}>n/a</option>
      }
      return(
        <option key={index} value={valenceOption}>{valenceOption * 10}</option>
      )
    })
  
    const mappedTargetTempo = options.map((tempoOption, index) => {
      if (tempoOption === '') {
        return <option key={index} value={''}>n/a</option>
      }
      return(
        <option key={index} value={tempoOption}>{tempoOption * 10}</option>
      )
    })
  
    const mappedTargetPopularity = options.map((popularityOption, index) => {
      if (popularityOption === '') {
        return <option key={index} value={''}>n/a</option>
      }
      return(
        <option key={index} value={popularityOption * 100}>{popularityOption * 10}</option>
      )
    })

  return (
    <div>
      <h3>Select Playlist Values Manually:</h3>
      <form>
        <label htmlFor="targetEnergy">Select Target Energy:</label>
        <select value={props.targetEnergy} name="targetEnergy" onChange={event => props.setEnergy(event.target.value)}>{mappedTargetEnergy}</select>
        <label htmlFor="targetValence">Select Target Happiness:</label>
        <select value={props.targetValence} name="targetValence" onChange={event => props.setValence(event.target.value)}>{mappedTargetValence}</select>
        <label htmlFor="targetTempo">Select Target Tempo:</label>
        <select value={props.targetTempo} name="targetTempo" onChange={event => props.setTempo(event.target.value)}>{mappedTargetTempo}</select>
        <label htmlFor="targetPopularity">Select Target Popularity:</label>
        <select value={props.targetPopularity} name="targetPopularity" onChange={event => props.setPopularity(event.target.value)}>{mappedTargetPopularity}</select>
      </form>
    </div>
  );
}

export default ChangeParams;