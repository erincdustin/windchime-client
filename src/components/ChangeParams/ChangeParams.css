.change-params {
  display: inline-block;
  text-align: center;
}
.param {
  text-align: left;
  padding: 12px;
  display: inline-block;
  border-radius: 15px;
  width: 90%;
  height: 25px;
  background-color: rgb(215, 240, 243);
  margin: 10px auto;
}

select{
  float: right;
  height: 25px;
  border: none;
  border-radius: 10px;
}

.center {
  margin-top: 10px;
  display: inline-block;
  text-align: center;
}

@media screen and (min-width: 650px) {
  .param {
    width: 450px;
  }
}