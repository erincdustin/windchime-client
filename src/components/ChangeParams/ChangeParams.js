import React from 'react';
import WindChimeContext from '../../contexts/windchime-context';
import WindChimeApiService from '../../services/api-service';
import TokenService from '../../services/token-service';
import './ChangeParams.css';

class ChangeParams extends React.Component {

  static contextType = WindChimeContext;

  handleGenrePlaylist = (genre) => {
    this.context.setTopArtists(null);
    WindChimeApiService.getGenreTracks(genre, this.context.targetEnergy, this.context.targetValence, this.context.targetTempo, this.context.targetPopularity)
      .then(res => {
        this.context.setSongs(res.tracks)
        const mappedSongs = res.tracks.map(song => {
          const songId = song.id;
          return `spotify:track:${songId}`;
          }).join(',');

        WindChimeApiService.makeGenrePlaylist(this.context.id, genre, this.context.weather)
          .then(res => {
            this.context.setPlaylistId(res.id)
            const playlistId = res.id

            WindChimeApiService.setSongs(res.id, mappedSongs)
              .then(res => {
                this.context.setSnapshot(res.snapshot_id)
                const playlistBody= JSON.stringify({
                  playlist_id: playlistId,
                  user_id: this.context.id,
                  energy: this.context.targetEnergy,
                  valence: this.context.targetValence,
                  tempo: this.context.targetTempo,
                  popularity: this.context.targetPopularity
                });

                WindChimeApiService.savePlaylist(playlistBody)
                  .then(res => {
                    console.log(`User playlist ${res.playlist_id} added`);
                    TokenService.savePlaylistToken(res.playlist_id);
                  })
              })
          })
      })
      .catch(err => {
        this.context.setError(err)
      })
  }

  handleArtistPlaylist = () => {
    this.context.setGenreChoice(null);
    WindChimeApiService.getUserArtists()
      .then(res => {
        const artists = res.items;
        let artistString = '';
        if(artists.length === 0) {
          artistString = '3WrFJ7ztbogyGnTHbHJFl2,08GQAI4eElDnROBrJRGE0X,0ECwFtbIWEVNwjlrfc6xoL'
        } else {
          artistString = artists.map(artist => {
            return artist.id;
          }).join(',');
        }
        this.context.setTopArtists(artistString)

      WindChimeApiService.getArtistTracks(artistString, this.context.targetEnergy, this.context.targetValence, this.context.targetTempo, this.context.targetPopularity)
        .then(res => {
          this.context.setSongs(res.tracks)
          const mappedSongs = res.tracks.map(song => {
            const songId = song.id;
            return `spotify:track:${songId}`;
            }).join(',');

          WindChimeApiService.makeArtistsPlaylist(this.context.id, this.context.weather)
            .then(res => {
              this.context.setPlaylistId(res.id)
              const playlistId = res.id

                WindChimeApiService.setSongs(res.id, mappedSongs)
                  .then(res => {
                    this.context.setSnapshot(res.snapshot_id)
                    const playlistBody= JSON.stringify({
                      playlist_id: playlistId,
                      user_id: this.context.id,
                      energy: this.context.targetEnergy,
                      valence: this.context.targetValence,
                      tempo: this.context.targetTempo,
                      popularity: this.context.targetPopularity
                    });

                    WindChimeApiService.savePlaylist(playlistBody)
                      .then(res => {
                        console.log(`User playlist ${res.playlist_id} added`);
                        TokenService.savePlaylistToken(res.playlist_id);
                      })
                })
            })
        })
      })
      .catch(err => {
        this.context.setError(err)
      })
  }

  handleSetEnergy = (energyLevel) => {
    if (energyLevel !== null) {
      this.context.setTargetEnergy(Number(energyLevel))
    } else {
      this.context.setTargetEnergy(null)
    }
  }

  handleSetValence = (valenceLevel) => {
    if (valenceLevel !== null) {
      this.context.setTargetValence(Number(valenceLevel))
    } else {
      this.context.setTargetValence(null)
    }
  }

  handleSetTempo = (tempoLevel) => {
    if (tempoLevel !== null) {
      this.context.setTargetTempo(Number(tempoLevel))
    } else {
      this.context.setTargetTempo(null)
    }
  }

  handleSetPopularity = (popularityLevel) => {
    if (popularityLevel !== null) {
      this.context.setTargetPopularity(Number(popularityLevel))
    } else {
      this.context.setTargetPopularity(null)
    }
  }

  mapValues = (value) => {
    const options = [null,.1,.2,.3,.4,.5,.6,.7,.8,.9,1];
    let mappedOptions = options.map((option, index) => {
      if (!option) {
        return <option key={index} value={null}>n/a</option>
      } else {
        if(value === 'popularity'){
          return <option key={index} value={option * 100}>{option * 10}</option>
        } else {
          return <option key={index} value={option}>{option * 10}</option>
        }
      }
    })
    return mappedOptions
  }

  render() {
    return (
      <div className="change-params">
        <h4>Select Playlist Values Manually:</h4>
        <form>
          <span className="param">
          <label htmlFor="targetEnergy"><h4 className="align">Select Target Energy:</h4></label>
          <select 
            value={Math.floor(this.context.targetEnergy*10)/10} 
            name="targetEnergy" 
            onChange={event => this.handleSetEnergy(event.target.value)}>{this.mapValues('energy')}
          </select>
          </span>
          <span className="param">
          <label htmlFor="targetValence"><h4 className="align">Select Target Happiness:</h4></label>
          <select 
            value={Math.floor(this.context.targetValence*10)/10} 
            name="targetValence" 
            onChange={event => this.handleSetValence(event.target.value)}>{this.mapValues('valence')}
          </select>
          </span>
          <span className="param">
          <label htmlFor="targetTempo"><h4 className="align">Select Target Tempo:</h4></label>
          <select 
            value={Math.floor(this.context.targetTempo*10)/10} 
            name="targetTempo" 
            onChange={event => this.handleSetTempo(event.target.value)}>{this.mapValues('tempo')}
          </select>
          </span>
          <span className="param">
          <label htmlFor="targetPopularity"><h4 className="align">Select Target Popularity:</h4></label>
          <select 
            value={Math.floor(this.context.targetPopularity*10)/10} 
            name="targetPopularity" 
            onChange={event => this.handleSetPopularity(event.target.value)}>{this.mapValues('popularity')}
          </select>
          </span>
        </form>
        <button className="btn center" 
          onClick={()=> this.props.history.goBack()}>Back</button>
        <button className="btn center" 
          onClick={()=> {
            if(!this.context.topArtists){
            this.handleGenrePlaylist(this.context.genreChoice);
            this.props.history.push('/results');
            }
            if(this.context.topArtists){
            this.handleArtistPlaylist(); 
            this.props.history.push('/results')
            }
          }}>Update</button>
      </div>
    );
  }
}

export default ChangeParams;