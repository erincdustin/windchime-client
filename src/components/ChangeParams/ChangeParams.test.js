import React from 'react';
import renderer from 'react-test-renderer';
import { library } from '@fortawesome/fontawesome-svg-core'
import { faWind } from '@fortawesome/free-solid-svg-icons'
import ChangeParams from './ChangeParams';

library.add(faWind)

describe(`ChangeParams`, () => {
  const props = {
    targetEnergy: .5,
    targetValence: .2,
    targetTempo: .3,
    targetPopularity: null,
    topArtists: [1,2,3,4,5],
    genreOption: 'folk',
    setEnergy: jest.fn(),
    setValence: jest.fn(),
    setTempo: jest.fn(),
    setPopularity: jest.fn(),
    getGenrePlaylist: jest.fn(),
    getArtistPlaylist: jest.fn(),
  }

  it('renders as expected', () => {
    const tree = renderer
      .create(<ChangeParams 
        setEnergy={props.setEnergy}
        targetEnergy={props.targetEnergy}
        setValence={props.setValence}
        targetValence={props.targetValence}
        setTempo={props.setTempo}
        targetTempo={props.targetTempo}
        setPopularity={props.setPopularity}
        targetPopularity={props.targetPopularity}
        getGenrePlaylist={props.getGenrePlaylist}
        getArtistPlaylist={props.getArtistPlaylist}
        topArtists={props.topArtists}
        genreOption={props.genreOption}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})
