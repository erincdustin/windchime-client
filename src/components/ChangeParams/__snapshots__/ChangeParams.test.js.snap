// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeParams renders as expected 1`] = `
<div
  className="change-params"
>
  <h4>
    Select Playlist Values Manually:
  </h4>
  <form>
    <span
      className="param"
    >
      <label
        htmlFor="targetEnergy"
      >
        <h4
          className="align"
        >
          Select Target Energy:
        </h4>
      </label>
      <select
        name="targetEnergy"
        onChange={[Function]}
        value={0}
      >
        <option
          value={null}
        >
          n/a
        </option>
        <option
          value={0.1}
        >
          1
        </option>
        <option
          value={0.2}
        >
          2
        </option>
        <option
          value={0.3}
        >
          3
        </option>
        <option
          value={0.4}
        >
          4
        </option>
        <option
          value={0.5}
        >
          5
        </option>
        <option
          value={0.6}
        >
          6
        </option>
        <option
          value={0.7}
        >
          7
        </option>
        <option
          value={0.8}
        >
          8
        </option>
        <option
          value={0.9}
        >
          9
        </option>
        <option
          value={1}
        >
          10
        </option>
      </select>
    </span>
    <span
      className="param"
    >
      <label
        htmlFor="targetValence"
      >
        <h4
          className="align"
        >
          Select Target Happiness:
        </h4>
      </label>
      <select
        name="targetValence"
        onChange={[Function]}
        value={0}
      >
        <option
          value={null}
        >
          n/a
        </option>
        <option
          value={0.1}
        >
          1
        </option>
        <option
          value={0.2}
        >
          2
        </option>
        <option
          value={0.3}
        >
          3
        </option>
        <option
          value={0.4}
        >
          4
        </option>
        <option
          value={0.5}
        >
          5
        </option>
        <option
          value={0.6}
        >
          6
        </option>
        <option
          value={0.7}
        >
          7
        </option>
        <option
          value={0.8}
        >
          8
        </option>
        <option
          value={0.9}
        >
          9
        </option>
        <option
          value={1}
        >
          10
        </option>
      </select>
    </span>
    <span
      className="param"
    >
      <label
        htmlFor="targetTempo"
      >
        <h4
          className="align"
        >
          Select Target Tempo:
        </h4>
      </label>
      <select
        name="targetTempo"
        onChange={[Function]}
        value={0}
      >
        <option
          value={null}
        >
          n/a
        </option>
        <option
          value={0.1}
        >
          1
        </option>
        <option
          value={0.2}
        >
          2
        </option>
        <option
          value={0.3}
        >
          3
        </option>
        <option
          value={0.4}
        >
          4
        </option>
        <option
          value={0.5}
        >
          5
        </option>
        <option
          value={0.6}
        >
          6
        </option>
        <option
          value={0.7}
        >
          7
        </option>
        <option
          value={0.8}
        >
          8
        </option>
        <option
          value={0.9}
        >
          9
        </option>
        <option
          value={1}
        >
          10
        </option>
      </select>
    </span>
    <span
      className="param"
    >
      <label
        htmlFor="targetPopularity"
      >
        <h4
          className="align"
        >
          Select Target Popularity:
        </h4>
      </label>
      <select
        name="targetPopularity"
        onChange={[Function]}
        value={0}
      >
        <option
          value={null}
        >
          n/a
        </option>
        <option
          value={10}
        >
          1
        </option>
        <option
          value={20}
        >
          2
        </option>
        <option
          value={30}
        >
          3
        </option>
        <option
          value={40}
        >
          4
        </option>
        <option
          value={50}
        >
          5
        </option>
        <option
          value={60}
        >
          6
        </option>
        <option
          value={70}
        >
          7
        </option>
        <option
          value={80}
        >
          8
        </option>
        <option
          value={90}
        >
          9
        </option>
        <option
          value={100}
        >
          10
        </option>
      </select>
    </span>
  </form>
  <button
    className="btn center"
    onClick={[Function]}
  >
    Back
  </button>
  <button
    className="btn center"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;
